payment/controllers.go

```diff
@@ -647,6 +694,97 @@ func MerchantTransactions(service *Service) handlers.AppHandler {
 	})
 }
 
+// HandleStripeWebhook is the handler for stripe checkout session webhooks
+func HandleStripeWebhook(service *Service) handlers.AppHandler {
+	return handlers.AppHandler(func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		// os.Getenv("STRIPE_SECRET")
+		stripe.Key = "sk_test_51HlmudHof20bphG6m8eJi9BvbPMLkMX4HPqLIiHmjdKAX21oJeO3S6izMrYTmiJm3NORBzUK1oM8STqClDRT3xQ700vyUyabNo"
+		// os.Getenv("STRIPE_WEBHOOK_SECRET")
+		endpointSecret := "whsec_Nm4yLVIpnG2cW5fW3kHQHxXBAJCL9dUj"
+
+		const MaxBodyBytes = int64(65536)
+		r.Body = http.MaxBytesReader(w, r.Body, MaxBodyBytes)
+
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			return handlers.WrapError(err, "error reading request body", http.StatusServiceUnavailable)
+		}
+
+		event, err := webhook.ConstructEvent(body, r.Header.Get("Stripe-Signature"), endpointSecret)
+		if err != nil {
+			return handlers.WrapError(err, "error verifying webhook signature", http.StatusBadRequest)
+		}
+
+		// Handle invoice events
+		if event.Type == "invoice.updated" {
+
+			// Retrieve invoice from update events
+			var invoice stripe.Invoice
+			err := json.Unmarshal(event.Data.Raw, &invoice)
+			if err != nil {
+				return handlers.WrapError(err, "error parsing webhook JSON", http.StatusBadRequest)
+			}
+
+			// Get the subscription and orderID connected to this invoice
+			sc := &client.API{}
+			// os.Getenv("STRIPE_SECRET")
+			sc.Init("sk_test_51HlmudHof20bphG6m8eJi9BvbPMLkMX4HPqLIiHmjdKAX21oJeO3S6izMrYTmiJm3NORBzUK1oM8STqClDRT3xQ700vyUyabNo", nil)
+			subscription, err := sc.Subscriptions.Get(invoice.Subscription.ID, nil)
+			if err != nil {
+				return handlers.WrapError(err, "error retrieving subscription", http.StatusInternalServerError)
+			}
+			orderID, err := uuid.FromString(subscription.Metadata["orderID"])
```

Comment:
should we be fully trusting the metadata on the subscription? or should we ensure that the subscription id / order correspond in our db?